addons/stock/tests/test_stock_return_picking.py

```diff
@@ -209,3 +209,67 @@ def test_stock_return_for_exchange(self):
         return_picking = self.env['stock.picking'].browse(res['res_id'])
         self.assertTrue(return_picking)
         self.assertEqual(len(return_picking.move_ids), 1)
+
+    def test_stock_picking_report_has_return(self):
+        """
+        Ensures that only returned serialized products are marked as returned.
+
+        Scenario:
+        - A delivery of two serialized units
+        - One unit is returned
+
+        Expected:
+        - The stock lot report lists two entries
+        - Only the returned unit has `has_return = True`, the other remains `False`
+        """
+        wh_stock = self.env['stock.location'].browse(self.stock_location)
+        partner = self.env['res.partner'].create({'name': 'Test Customer'})
+        quant = self.env['stock.quant']
+
+        product_serial = self.env['product.product'].create({
+            'name': 'Tracked by SN',
+            'is_storable': True,
+            'tracking': 'serial',
+        })
+
+        serial_1 = self.env['stock.lot'].create({'name': 'SN1', 'product_id': product_serial.id})
+        serial_2 = self.env['stock.lot'].create({'name': 'SN2', 'product_id': product_serial.id})
+
+        quant._update_available_quantity(product_serial, wh_stock, 1.0, lot_id=serial_1)
+        quant._update_available_quantity(product_serial, wh_stock, 1.0, lot_id=serial_2)
+
+        picking = self.PickingObj.create({
+            'partner_id': partner.id,
+            'picking_type_id': self.picking_type_out,
+            'location_id': self.stock_location,
+            'location_dest_id': self.customer_location,
+            'move_ids': [(0, 0, {
+                'name': 'Move SN',
+                'product_id': product_serial.id,
+                'product_uom_qty': 2,
+                'product_uom': product_serial.uom_id.id,
+                'location_id': self.stock_location,
+                'location_dest_id': self.customer_location,
+            })],
+        })
+
+        picking.action_confirm()
+        picking.action_assign()
+        picking.move_ids.picked = True
+        picking.button_validate()
+
+        return_wizard = self.env['stock.return.picking'].with_context(active_id=picking.id, active_model='stock.picking').create({})
+        return_wizard.product_return_moves.quantity = 1
+        res = return_wizard.action_create_returns()
+        return_picking = self.PickingObj.browse(res["res_id"])
+
+        return_picking.action_confirm()
+        return_picking.move_ids.picked = True
+        return_picking.button_validate()
+        self.env.flush_all()
+
+        lot_report = self.env['stock.lot.report'].search([
+            ('partner_id', '=', partner.id),
+        ])
+        lot_returns = {record.lot_id.id: record.has_return for record in lot_report}
+        self.assertTrue(lot_returns[serial_1.id] != lot_returns[serial_2.id])
```

Comment:
It's correct
It's a bit weird `assertTrue(... != ...)`
I would rather go for either two assert (`assertTrue` and `assertFalse`), or still the `assertRecordValues`, which I think is the more clear approach (so we get rid of the dict creation)

```suggestion
        ], order='id')
        self.assertRecordValues(lot_report, [
            {'lot_id': serial_1.id, 'has_return': True},
            {'lot_id': serial_2.id, 'has_return': False},
        ])
```

Reply:
@adwid 
I thought this way could solve the possible non-deterministic issue. I currently encounter no problem with your implementation so I will apply it but wouldn't it be safer to totally avoid non-deterministic flows ?

Reply:
That's why I added the `order='id'`, which refers to SML ID :)